vm/src/stdlib/signal.rs

```diff
@@ -0,0 +1,165 @@
+use crate::vm::VirtualMachine;
+
+use std::sync::atomic::{AtomicBool, Ordering};
+
+use num_traits::cast::ToPrimitive;
+
+use arr_macro::arr;
+
+#[cfg(unix)]
+use nix::unistd::alarm as sig_alarm;
+
+use libc;
+
+#[cfg(not(windows))]
+use libc::{SIG_DFL, SIG_ERR, SIG_IGN};
+
+#[cfg(windows)]
+const SIG_DFL: libc::sighandler_t = 0;
+#[cfg(windows)]
+const SIG_IGN: libc::sighandler_t = 1;
+#[cfg(windows)]
+const SIG_ERR: libc::sighandler_t = !0;
+
+const NSIG: usize = 64;
+
+// We cannot use the NSIG const in the arr macro. This will fail compilation if NSIG is different.
+static mut TRIGGERS: [AtomicBool; NSIG] = arr![AtomicBool::new(false); 64];
+
+extern "C" fn run_signal(signum: i32) {
+    unsafe {
+        TRIGGERS[signum as usize].store(true, Ordering::Relaxed);
+    }
+}
+
+fn signal(
+    signalnum: PyIntRef,
+    handler: PyObjectRef,
+    vm: &VirtualMachine,
+) -> PyResult<Option<PyObjectRef>> {
+    if !vm.isinstance(&handler, &vm.ctx.function_type())?
+        && !vm.isinstance(&handler, &vm.ctx.bound_method_type())?
+        && !vm.isinstance(&handler, &vm.ctx.builtin_function_or_method_type())?
+    {
+        return Err(vm.new_type_error("Hanlder must be callable".to_string()));
+    }
+    let signal_module = vm.import("signal", &vm.ctx.new_tuple(vec![]), 0)?;
+    let sig_dfl = vm.get_attribute(signal_module.clone(), "SIG_DFL")?;
+    let sig_ign = vm.get_attribute(signal_module, "SIG_IGN")?;
+    let signalnum = signalnum.as_bigint().to_i32().unwrap();
+    check_signals(vm);
+    let sig_handler = if handler.is(&sig_dfl) {
+        SIG_DFL
+    } else if handler.is(&sig_ign) {
+        SIG_IGN
+    } else {
+        run_signal as libc::sighandler_t
+    };
+    let old = unsafe { libc::signal(signalnum, sig_handler) };
+    if old == SIG_ERR {
+        return Err(vm.new_os_error("Failed to set signal".to_string()));
+    }
+    let old_handler = vm.signal_handlers.borrow_mut().insert(signalnum, handler);
+    Ok(old_handler)
+}
+
+fn getsignal(signalnum: PyIntRef, vm: &VirtualMachine) -> PyResult<Option<PyObjectRef>> {
+    let signalnum = signalnum.as_bigint().to_i32().unwrap();
+    Ok(vm.signal_handlers.borrow_mut().get(&signalnum).cloned())
+}
+
+#[cfg(unix)]
+fn alarm(time: PyIntRef, _vm: &VirtualMachine) -> u32 {
+    let time = time.as_bigint().to_u32().unwrap();
+    let prev_time = if time == 0 {
+        sig_alarm::cancel()
+    } else {
+        sig_alarm::set(time)
+    };
+    prev_time.unwrap_or(0)
+}
+
+#[allow(clippy::needless_range_loop)]
+pub fn check_signals(vm: &VirtualMachine) {
+    for signum in 1..NSIG {
+        let triggerd = unsafe { TRIGGERS[signum].swap(false, Ordering::Relaxed) };
+        if triggerd {
+            let handler = vm
+                .signal_handlers
+                .borrow()
+                .get(&(signum as i32))
+                .expect("Handler should be set")
+                .clone();
+            vm.invoke(handler, vec![vm.new_int(signum), vm.get_none()])
```

Comment:
What should happen when a signal handler throws an exception? Do you want to tackle this now, or implement it later?

Reply:
I would prefer to do it later